### ISSUE TITLE
Make i18n module work with url-rewrite

### ISSUE THREAD

**Issue by zhendalf** [nuxt-modules/i18n#915]
### Version
nuxt-i18n:  6.15.1
nuxt: 2.14.6

#### mode:
 - [x] static

Hi. I am generating a static website in multiple languages using Nuxt. The site is hosted on Firebase Hosting, which now provides [i18n rewrites](https://firebase.google.com/docs/hosting/i18n-rewrites). It means that if a spanish-speaking visitor visits my site, I can serve content from `/es/` without doing a redirect. 

The rewrite part works great, but the problem is that as soon javascript kicks in, the page gets translated into default locale. How can I make i18n module agree with the page's language, even if it is not under localized path in the URL?

My hunch is that it needs to look into `__NUXT__.routePath` instead of looking at the URL.

### Nuxt-i18n configuration

```
i18n: {
  detectBrowserLanguage: false, // Avoid redirect
  locales: ['en', 'fr', 'es', 'ru'],
  defaultLocale: 'fr',
  vueI18n: {
    fallbackLocale: 'fr',
    messages: {
      en: {
        welcome: 'Welcome',
      },
      fr: {
        welcome: 'Bienvenue',
      },
      es: {
        welcome: 'Bienvenido',
      },
      ru: {
        welcome: 'Привет',
      },
    },
  },
}
```

### Reproduction Link

https://eh-test-i18n.web.app/

### Steps to reproduce

1. Visit the link above
2. If your language is english, you will first see 'Welcome' (served from `/en` thanks to hosting rewrite)
3. When javascript hydrates, the language will be switched to default French ('Welcome' will be replaced with 'Bienvenue')

### What is Expected?

I expect the language of the page to stay in english and not be switched to French. I want to keep the page static, without adding server-based redirects or client-based redirects.

Here is a repo with demo's source code: https://github.com/ebeloded/nuxt-i18n-url-rewrite-test

**Comment by rchl** [comment]
That sounds like a big hack and not SEO friendly at all (you maybe don't care about that). I don't think I would want to go into this area as it seems like a can of worms (there might be places in the module that work based on the location). If you will figure out something that looks clean then feel free to create a PR. If the solution won't be clean then you might have to create and use your own fork for that.

**Comment by rchl** [comment]
Just some more info: Maybe `nuxt-i18n` doesn't rely on paths itself but `VueRouter` routing is based on paths so even without this module, you would likely have issues with that approach (if let's say, you would create `index.vue` and `fr/index.vue` and serve the `fr/index.vue` for french).

**Comment by zhendalf** [comment]
@rchl Thanks for your responses
SEO is important, as well as user experience, which is why I am trying to integrate the rewrite. A language-based URL rewrite is not an esoteric practice. For example, https://google.com will be displayed to you in your language by default, without a redirect, so do many other well-designed multi-language websites.

I was hoping there would be a solution. If not - other people may be coming with the same question, as i18n rewrite is a great feature of Firebase Hosting.

**Comment by zhendalf** [comment]
Regarding a possible fix: is there a place where I would be able to execute `history.replaceState(null, null, __NUXT__.routePath)` before anything else is kicked into gear?

This way the URL should satisfy everyone down the line and the content would be served in the right language.

**Comment by rchl** [comment]
I doubt this will be a reliable solution since routes are created at build time and have paths embedded in them. So, for example, all links to other routes would still be wrong. I think that what you are trying to do is not very Nuxt-friendly and I think the issue is not specific to this module.

You can experiment with pure Nuxt to see if you can make it work. And if you can, then you can add `nuxt-i18n` and see what extra issues it poses. Making everything work correctly with just Nuxt will be hard if not impossible, IMO.

**Comment by rchl** [comment]
> SEO is important, as well as user experience, which is why I am trying to integrate the rewrite.

If SEO is important then how would that work with crawlers? Wouldn't crawlers always get a version of the page in a specific locale (likely English) and not be able to access other locales?

**Comment by zhendalf** [comment]
I have figured out how to make it work. I modified the nuxt template and added a line to replace history state:
```html
<!DOCTYPE html>
<html {{ HTML_ATTRS }}>
  <head {{ HEAD_ATTRS }}>
    {{ HEAD }}
  </head>
  <body {{ BODY_ATTRS }}>
    {{ APP }}
    <script>
      history.replaceState(null, null, window.__NUXT__.routePath)
    </script>
  </body>
</html>
```

When a user is served a page via URL rewrite, the URL is changed before nuxt/vue-router/i18n have had access to it. As a result, both the page content and the page URL look as if the user was redirected to rewrite destination.

This allows serving statically generated content from CDN in user's language, without a redirect 🎉🕺🏼

It is a solution to my original question, so this issue can be closed. 

There is, however, a SEO-related issue that emerges from default expectations of i18n-module.

@rchl, to make this work, one would have to make a different use of [rel=canonical](https://support.google.com/webmasters/answer/139066?hl=en) and [x-default](https://support.google.com/webmasters/answer/189077?hl=en#xdefault). 

I'll explain in a separate issue to keep things to a point.
